go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -1323,55 +1316,61 @@ func (npw *nodePortWatcher) AddEndpointSlice(epSlice *discovery.EndpointSlice) e
 			epSlice.Name, epSlice.Namespace)
 		return nil
 	}
-
 	if !util.ServiceTypeHasClusterIP(svc) || !util.IsClusterIPSet(svc) {
 		return nil
 	}
 
 	klog.V(5).Infof("Adding endpointslice %s in namespace %s", epSlice.Name, epSlice.Namespace)
-	nodeIPs, _ := npw.nodeIPManager.ListAddresses()
 	epSlices, err := npw.watchFactory.GetServiceEndpointSlices(svc.Namespace, svc.Name, netInfo.GetNetworkName())
 	if err != nil {
 		// No need to continue adding the new endpoint slice, if we can't retrieve all slices for this service
 		return fmt.Errorf("error retrieving endpointslices for service %s/%s during endpointslice add: %w", svc.Namespace, svc.Name, err)
 	}
-	localEndpoints := npw.GetLocalEligibleEndpointAddresses(epSlices, svc)
-	hasLocalHostNetworkEp := util.HasLocalHostNetworkEndpoints(localEndpoints, nodeIPs)
+
+	var errors []error
+	nodeIPs, _ := npw.nodeIPManager.ListAddresses()
+	localHostNetworkEndpoints, localEndpoints := npw.GetLocalEligibleEndpointAddresses(epSlices, svc).ExtractHostNetworkEndpoints(nodeIPs)
 
 	// Here we make sure the correct rules are programmed whenever an AddEndpointSlice event is
 	// received, only alter flows if we need to, i.e if cache wasn't set or if it was and
-	// hasLocalHostNetworkEp or localEndpoints state (for LB svc where NPs=0) changed, to prevent flow churn
+	// localHostNetworkEndpoints or localEndpoints state (for LB svc where NPs=0) changed, to prevent flow churn.
 	out, exists := npw.getServiceInfo(*svcNamespacedName)
 	if !exists {
 		klog.V(5).Infof("Endpointslice %s ADD event in namespace %s is creating rules", epSlice.Name, epSlice.Namespace)
-		if err = addServiceRules(svc, netInfo, localEndpoints, hasLocalHostNetworkEp, npw); err != nil {
+		if err := addServiceRules(svc, netInfo, localHostNetworkEndpoints, localEndpoints, npw); err != nil {
```

Comment:
todo: this line here isn't covered by unit tests, nor is pw.addOrSetServiceInfo below